pennylane/transforms/diagonalize_measurements.py

```diff
@@ -0,0 +1,175 @@
+"""Transform to diagonalize measurements on a tape, assuming all measurements are commuting."""
+
+from copy import copy
+from functools import singledispatch
+
+import pennylane as qml
+from pennylane.devices.preprocess import null_postprocessing
+from pennylane.operation import Tensor
+from pennylane.ops import CompositeOp, SymbolicOp
+from pennylane.transforms.core import transform
+
+# pylint: disable=protected-access
+
+_default_unsupported_obs = frozenset(["PauliX", "PauliY", "Hadamard"])
+
+
+@transform
+def diagonalize_tape_measurements(tape, unsupported_obs=_default_unsupported_obs):
+    """Diagonalize the measurements on a tape if they are not supported. Assumes all observables commute."""
```

Comment:
What would happen if the circuit contains non-commuting observables on the same wire? Would the error message be easy to understand?

Reply:
I'm not even sure there would be an error message in its current form, I think it would just do something messy and wrong. We'll add a check and raise an error :)

Reply:
Now it raises an error